client/amqp/v2/amqp.go

```diff
@@ -72,14 +91,16 @@ func (tc *Publisher) Publish(ctx context.Context, exchange, key string, mandator
 	}
 
 	c := amqpHeadersCarrier(msg.Headers)
-	err := sp.Tracer().Inject(sp.Context(), opentracing.TextMap, c)
-	if err != nil {
-		return fmt.Errorf("failed to inject tracing headers: %w", err)
+
+	if err := sp.Tracer().Inject(sp.Context(), opentracing.TextMap, c); err != nil {
+		log.FromContext(ctx).Errorf("failed to inject tracing headers: %v", err)
 	}
 	msg.Headers[correlation.HeaderID] = correlation.IDFromContext(ctx)
 
-	err = tc.channel.Publish(exchange, key, mandatory, immediate, msg)
-	trace.SpanComplete(sp, err)
+	start := time.Now()
+	err := tc.channel.Publish(exchange, key, mandatory, immediate, msg)
+
+	observePublish(sp, start, exchange, err)
```

Comment:
would it be a more consistent approach to defer this logic right after we start the span ?